cpp/src/arrow/array.h

```diff
@@ -274,23 +274,28 @@ class ARROW_EXPORT Array {
   const uint8_t* null_bitmap_data() const { return null_bitmap_data_; }
 
   /// Equality comparison with another array
-  bool Equals(const Array& arr) const;
-  bool Equals(const std::shared_ptr<Array>& arr) const;
+  bool Equals(const Array& arr, bool nans_equal = false) const;
+  bool Equals(const std::shared_ptr<Array>& arr, bool nans_equal = false) const;
```

Comment:
It's a bit gnarly to have this in the API for all Array subclasses, but I'm not sure what is a better alternative (e.g. a `ComparisonOptions` struct of some kind)